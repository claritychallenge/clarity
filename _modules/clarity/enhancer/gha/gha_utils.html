

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../../../">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>clarity.enhancer.gha.gha_utils &mdash; pyClarity 0.7.1.post21 documentation</title>
      <link rel="stylesheet" type="text/css" href="../../../../_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="../../../../_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="../../../../_static/custom.css" />

  
      <script src="../../../../_static/jquery.js?v=5d32c60e"></script>
      <script src="../../../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../../../../_static/documentation_options.js?v=5cc28f93"></script>
      <script src="../../../../_static/doctools.js?v=9a2dae69"></script>
      <script src="../../../../_static/sphinx_highlight.js?v=dc90522c"></script>
      <script type="module" src="https://cdn.jsdelivr.net/npm/mermaid@11.2.0/dist/mermaid.esm.min.mjs"></script>
      <script type="module" src="https://cdn.jsdelivr.net/npm/@mermaid-js/layout-elk@0.1.4/dist/mermaid-layout-elk.esm.min.mjs"></script>
      <script type="module">import mermaid from "https://cdn.jsdelivr.net/npm/mermaid@11.2.0/dist/mermaid.esm.min.mjs";import elkLayouts from "https://cdn.jsdelivr.net/npm/@mermaid-js/layout-elk@0.1.4/dist/mermaid-layout-elk.esm.min.mjs";mermaid.registerLayoutLoaders(elkLayouts);mermaid.initialize({startOnLoad:false});</script>
      <script src="https://cdn.jsdelivr.net/npm/d3@7.9.0/dist/d3.min.js"></script>
      <script type="module">
import mermaid from "https://cdn.jsdelivr.net/npm/mermaid@11.2.0/dist/mermaid.esm.min.mjs";
window.addEventListener("load", () => mermaid.run());
</script>
    <script src="../../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../../genindex.html" />
    <link rel="search" title="Search" href="../../../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../../index.html" class="icon icon-home">
            pyClarity
              <img src="../../../../_static/challenges.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../../install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../api.html">API Reference</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../../index.html">pyClarity</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../../index.html">Module code</a></li>
      <li class="breadcrumb-item active">clarity.enhancer.gha.gha_utils</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for clarity.enhancer.gha.gha_utils</h1><div class="highlight"><pre>
<span></span><span class="kn">from</span><span class="w"> </span><span class="nn">__future__</span><span class="w"> </span><span class="kn">import</span> <span class="n">annotations</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">logging</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">re</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">typing</span><span class="w"> </span><span class="kn">import</span> <span class="n">TYPE_CHECKING</span><span class="p">,</span> <span class="n">TypedDict</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">scipy.interpolate</span><span class="w"> </span><span class="kn">import</span> <span class="n">interp1d</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">clarity.enhancer.gha.gainrule_camfit</span><span class="w"> </span><span class="kn">import</span> <span class="n">gainrule_camfit_compr</span>

<span class="k">if</span> <span class="n">TYPE_CHECKING</span><span class="p">:</span>
    <span class="kn">from</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="kn">import</span> <span class="n">ndarray</span>

    <span class="kn">from</span><span class="w"> </span><span class="nn">clarity.utils.audiogram</span><span class="w"> </span><span class="kn">import</span> <span class="n">Audiogram</span>


<div class="viewcode-block" id="Gaintable">
<a class="viewcode-back" href="../../../../clarity.enhancer.gha.html#clarity.enhancer.gha.gha_utils.Gaintable">[docs]</a>
<span class="k">class</span><span class="w"> </span><span class="nc">Gaintable</span><span class="p">(</span><span class="n">TypedDict</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Gaintable for a given audiogram.&quot;&quot;&quot;</span>

    <span class="n">sGt_uncorr</span><span class="p">:</span> <span class="n">ndarray</span>
    <span class="n">sGt</span><span class="p">:</span> <span class="n">ndarray</span>
    <span class="n">noisegatelevel</span><span class="p">:</span> <span class="n">ndarray</span>
    <span class="n">noisegateslope</span><span class="p">:</span> <span class="n">ndarray</span>
    <span class="n">frequencies</span><span class="p">:</span> <span class="n">ndarray</span>
    <span class="n">levels</span><span class="p">:</span> <span class="n">ndarray</span>
    <span class="n">channels</span><span class="p">:</span> <span class="nb">int</span></div>



<div class="viewcode-block" id="FittingParams">
<a class="viewcode-back" href="../../../../clarity.enhancer.gha.html#clarity.enhancer.gha.gha_utils.FittingParams">[docs]</a>
<span class="k">class</span><span class="w"> </span><span class="nc">FittingParams</span><span class="p">(</span><span class="n">TypedDict</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Fitting parameters for gaintable calculation.&quot;&quot;&quot;</span>

    <span class="n">frequencies</span><span class="p">:</span> <span class="n">ndarray</span>
    <span class="n">edge_frequencies</span><span class="p">:</span> <span class="n">ndarray</span>
    <span class="n">levels</span><span class="p">:</span> <span class="n">ndarray</span>
    <span class="n">channels</span><span class="p">:</span> <span class="nb">int</span>
    <span class="n">side</span><span class="p">:</span> <span class="nb">str</span></div>



<div class="viewcode-block" id="get_gaintable">
<a class="viewcode-back" href="../../../../clarity.enhancer.gha.html#clarity.enhancer.gha.gha_utils.get_gaintable">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">get_gaintable</span><span class="p">(</span>
    <span class="n">audiogram_left</span><span class="p">:</span> <span class="n">Audiogram</span><span class="p">,</span>
    <span class="n">audiogram_right</span><span class="p">:</span> <span class="n">Audiogram</span><span class="p">,</span>
    <span class="n">noisegate_levels</span><span class="p">:</span> <span class="nb">float</span> <span class="o">|</span> <span class="n">ndarray</span><span class="p">,</span>
    <span class="n">noisegate_slope</span><span class="p">:</span> <span class="nb">float</span> <span class="o">|</span> <span class="n">ndarray</span><span class="p">,</span>
    <span class="n">cr_level</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span>
    <span class="n">max_output_level</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Gaintable</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Compute a gaintable for a given audiogram.</span>

<span class="sd">    Replaces MATLAB GUI interface of original OpenMHA software for</span>
<span class="sd">    gaintable_camfit_compr table calculation. Assumes two channels and</span>
<span class="sd">    that audiogram frequencies are identical at two ears.</span>

<span class="sd">    Args:</span>
<span class="sd">        audiogram_left (Audiogram): the audiogram for the left ear</span>
<span class="sd">        audiogram_right (Audiogram): the audiogram for the right ear</span>
<span class="sd">        audf (list): audiogram frequencies for fitting</span>
<span class="sd">        noisegatelevels (ndarray): compression threshold levels for each frequency band</span>
<span class="sd">        noisegateslope (float): determines slope of gains below compression threshold</span>
<span class="sd">        cr_level (int): overall input level in dB for calculation of compression ratios</span>
<span class="sd">        max_output_level (int): maximum output level in dB</span>

<span class="sd">    Returns:</span>
<span class="sd">        dict: dim ndarray of gains</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Fixed centre frequencies for amplification bands</span>
    <span class="n">sFitmodel</span><span class="p">:</span> <span class="n">FittingParams</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;frequencies&quot;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
            <span class="p">[</span><span class="mi">177</span><span class="p">,</span> <span class="mi">297</span><span class="p">,</span> <span class="mi">500</span><span class="p">,</span> <span class="mi">841</span><span class="p">,</span> <span class="mi">1414</span><span class="p">,</span> <span class="mi">2378</span><span class="p">,</span> <span class="mf">4.0000e03</span><span class="p">,</span> <span class="mf">6.7270e03</span><span class="p">,</span> <span class="mi">11314</span><span class="p">]</span>
        <span class="p">),</span>
        <span class="s2">&quot;edge_frequencies&quot;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
            <span class="p">[</span>
                <span class="mf">1.0000e-08</span><span class="p">,</span>
                <span class="mf">229.2793</span><span class="p">,</span>
                <span class="mf">385.3570</span><span class="p">,</span>
                <span class="mf">648.4597</span><span class="p">,</span>
                <span class="mf">1.0905e03</span><span class="p">,</span>
                <span class="mf">1.8337e03</span><span class="p">,</span>
                <span class="mf">3.0842e03</span><span class="p">,</span>
                <span class="mf">5.1873e03</span><span class="p">,</span>
                <span class="mf">8.7241e03</span><span class="p">,</span>
                <span class="mi">10000001</span><span class="p">,</span>
            <span class="p">]</span>
        <span class="p">),</span>
        <span class="s2">&quot;levels&quot;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="o">-</span><span class="mi">10</span><span class="p">,</span> <span class="mi">110</span> <span class="o">+</span> <span class="mi">1</span><span class="p">),</span>  <span class="c1"># Levels SPL at which to compute the gains</span>
        <span class="s2">&quot;channels&quot;</span><span class="p">:</span> <span class="mi">2</span><span class="p">,</span>
        <span class="s2">&quot;side&quot;</span><span class="p">:</span> <span class="s2">&quot;lr&quot;</span><span class="p">,</span>
    <span class="p">}</span>

    <span class="c1"># Calculate gains and compression ratios</span>
    <span class="n">gain_table</span><span class="p">,</span> <span class="n">noisegate_levels</span><span class="p">,</span> <span class="n">noisegate_slope</span> <span class="o">=</span> <span class="n">gainrule_camfit_compr</span><span class="p">(</span>
        <span class="n">audiogram_left</span><span class="p">,</span>
        <span class="n">audiogram_right</span><span class="p">,</span>
        <span class="n">sFitmodel</span><span class="p">,</span>
        <span class="n">noisegate_levels</span><span class="p">,</span>
        <span class="n">noisegate_slope</span><span class="p">,</span>
        <span class="n">cr_level</span><span class="p">,</span>
        <span class="n">max_output_level</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">gain_table_corrected</span> <span class="o">=</span> <span class="n">multifit_apply_noisegate</span><span class="p">(</span>
        <span class="n">gain_table</span><span class="p">,</span>
        <span class="n">sFitmodel</span><span class="p">[</span><span class="s2">&quot;frequencies&quot;</span><span class="p">],</span>
        <span class="n">sFitmodel</span><span class="p">[</span><span class="s2">&quot;levels&quot;</span><span class="p">],</span>
        <span class="n">noisegate_levels</span><span class="p">,</span>
        <span class="n">noisegate_slope</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="c1"># Reshape sGt here to suit cfg file input</span>
    <span class="n">sGt_data</span><span class="p">:</span> <span class="n">Gaintable</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;sGt_uncorr&quot;</span><span class="p">:</span> <span class="n">gain_table</span><span class="p">,</span>  <span class="c1"># sGt without noisegate correction</span>
        <span class="s2">&quot;sGt&quot;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">transpose</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">gain_table_corrected</span><span class="p">,</span> <span class="p">(</span><span class="mi">121</span><span class="p">,</span> <span class="mi">18</span><span class="p">),</span> <span class="n">order</span><span class="o">=</span><span class="s2">&quot;F&quot;</span><span class="p">)),</span>
        <span class="s2">&quot;noisegatelevel&quot;</span><span class="p">:</span> <span class="n">noisegate_levels</span><span class="p">,</span>
        <span class="s2">&quot;noisegateslope&quot;</span><span class="p">:</span> <span class="n">noisegate_slope</span><span class="p">,</span>
        <span class="s2">&quot;frequencies&quot;</span><span class="p">:</span> <span class="n">sFitmodel</span><span class="p">[</span><span class="s2">&quot;frequencies&quot;</span><span class="p">],</span>
        <span class="s2">&quot;levels&quot;</span><span class="p">:</span> <span class="n">sFitmodel</span><span class="p">[</span><span class="s2">&quot;levels&quot;</span><span class="p">],</span>
        <span class="s2">&quot;channels&quot;</span><span class="p">:</span> <span class="n">sFitmodel</span><span class="p">[</span><span class="s2">&quot;channels&quot;</span><span class="p">],</span>
    <span class="p">}</span>
    <span class="k">return</span> <span class="n">sGt_data</span></div>



<div class="viewcode-block" id="format_gaintable">
<a class="viewcode-back" href="../../../../clarity.enhancer.gha.html#clarity.enhancer.gha.gha_utils.format_gaintable">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">format_gaintable</span><span class="p">(</span>
    <span class="n">gaintable</span><span class="p">:</span> <span class="n">Gaintable</span><span class="p">,</span>
    <span class="n">noisegate_corr</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Format gaintable for insertion into cfg file as long string.</span>

<span class="sd">    Args:</span>
<span class="sd">        gaintable (ndarray): The gaintable to format</span>
<span class="sd">        noisegate_corr (boolean, optional): apply noisegate correction or do not</span>
<span class="sd">            (default: True)</span>

<span class="sd">    Returns:</span>
<span class="sd">        str: gaintable formatted for insertion into OpenMHA</span>
<span class="sd">            cfg file</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">noisegate_corr</span><span class="p">:</span>
        <span class="n">sGt</span> <span class="o">=</span> <span class="n">gaintable</span><span class="p">[</span><span class="s2">&quot;sGt&quot;</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;Noise gate correction not being applied to gain table.&quot;</span><span class="p">)</span>
        <span class="n">sGt</span> <span class="o">=</span> <span class="n">gaintable</span><span class="p">[</span><span class="s2">&quot;sGt_uncorr&quot;</span><span class="p">]</span>

    <span class="c1"># Do not apply gains to 9th and 18th row</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;No application of gains to 9th and 18th row of table.&quot;</span><span class="p">)</span>
    <span class="n">sGt</span><span class="p">[[</span><span class="mi">8</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="p">:]</span> <span class="o">=</span> <span class="mi">0</span>

    <span class="c1"># Format for inclusion in cfg file</span>
    <span class="n">v</span> <span class="o">=</span> <span class="s2">&quot;[&quot;</span>
    <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">shape</span><span class="p">(</span><span class="n">sGt</span><span class="p">)[</span><span class="mi">0</span><span class="p">]):</span>
        <span class="n">v</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">sGt</span><span class="p">[</span><span class="n">k</span><span class="p">,</span><span class="w"> </span><span class="p">:]</span><span class="si">}</span><span class="s2">;&quot;</span>
    <span class="n">v</span> <span class="o">+=</span> <span class="s2">&quot;]&quot;</span>
    <span class="n">v</span> <span class="o">=</span> <span class="n">v</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
    <span class="n">v</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="s2">&quot; +&quot;</span><span class="p">,</span> <span class="s2">&quot; &quot;</span><span class="p">,</span> <span class="n">v</span><span class="p">)</span>  <span class="c1"># remove extra white spaces</span>
    <span class="c1"># v = re.sub(&quot;. &quot;, &quot; &quot;, v)  # remove point after integers</span>

    <span class="n">formatted_sGt</span> <span class="o">=</span> <span class="n">v</span>

    <span class="k">return</span> <span class="n">formatted_sGt</span></div>



<div class="viewcode-block" id="multifit_apply_noisegate">
<a class="viewcode-back" href="../../../../clarity.enhancer.gha.html#clarity.enhancer.gha.gha_utils.multifit_apply_noisegate">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">multifit_apply_noisegate</span><span class="p">(</span>
    <span class="n">gain_table</span><span class="p">:</span> <span class="n">ndarray</span><span class="p">,</span>
    <span class="n">sFit_model_frequencies</span><span class="p">:</span> <span class="n">ndarray</span><span class="p">,</span>
    <span class="n">sFit_model_levels</span><span class="p">:</span> <span class="n">ndarray</span><span class="p">,</span>
    <span class="n">noisegate_level</span><span class="p">:</span> <span class="n">ndarray</span><span class="p">,</span>
    <span class="n">noisegate_slope</span><span class="p">:</span> <span class="n">ndarray</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">ndarray</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Apply noisegate to the gain table</span>

<span class="sd">    Based on OpenMHA subfunction of libmultifit.m.</span>

<span class="sd">    Args:</span>
<span class="sd">        gain_table (ndarray): gain array</span>
<span class="sd">        sFit_model_frequencies (list): FFT filterbank frequencies</span>
<span class="sd">        sFit_model_levels (ndarray): levels at which to calculate gains</span>
<span class="sd">        noisegate_level (ndarray): chosen compression threshold</span>
<span class="sd">        noisegate_slope (ndarray): determines slope below compression threshold</span>

<span class="sd">    Returns:</span>
<span class="sd">        ndarray: gain table with noisegate applied</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">gain_table_corrected</span> <span class="o">=</span> <span class="n">gain_table</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">]:</span>
        <span class="k">for</span> <span class="n">kf</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">sFit_model_frequencies</span><span class="p">)):</span>
            <span class="n">gain_noisegate</span> <span class="o">=</span> <span class="n">interp1d</span><span class="p">(</span>
                <span class="n">sFit_model_levels</span><span class="p">,</span>
                <span class="n">gain_table_corrected</span><span class="p">[:,</span> <span class="n">kf</span><span class="p">,</span> <span class="n">i</span><span class="p">],</span>
                <span class="n">fill_value</span><span class="o">=</span><span class="s2">&quot;extrapolate&quot;</span><span class="p">,</span>
            <span class="p">)(</span><span class="n">noisegate_level</span><span class="p">[</span><span class="n">kf</span><span class="p">,</span> <span class="n">i</span><span class="p">])</span>
            <span class="n">idx</span> <span class="o">=</span> <span class="p">[</span>
                <span class="n">i</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">sFit_model_levels</span> <span class="o">&lt;</span> <span class="n">noisegate_level</span><span class="p">[</span><span class="n">kf</span><span class="p">,</span> <span class="n">i</span><span class="p">])</span> <span class="k">if</span> <span class="n">x</span>
            <span class="p">]</span>
            <span class="n">gain_table_corrected</span><span class="p">[</span><span class="n">idx</span><span class="p">,</span> <span class="n">kf</span><span class="p">,</span> <span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
                <span class="n">sFit_model_levels</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span> <span class="o">-</span> <span class="n">noisegate_level</span><span class="p">[</span><span class="n">kf</span><span class="p">,</span> <span class="n">i</span><span class="p">]</span>
            <span class="p">)</span> <span class="o">*</span> <span class="n">noisegate_slope</span><span class="p">[</span><span class="n">kf</span><span class="p">,</span> <span class="n">i</span><span class="p">]</span> <span class="o">+</span> <span class="n">gain_noisegate</span>

    <span class="k">return</span> <span class="n">gain_table_corrected</span></div>

</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2020-2025, pyClarity authors.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>