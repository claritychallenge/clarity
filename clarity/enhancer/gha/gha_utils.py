import copy
import logging
import re

import numpy as np
from scipy.interpolate import interp1d

from clarity.enhancer.gha.gainrule_camfit import gainrule_camfit_compr


def get_gaintable(
    audiogram, noisegatelevels, noisegateslope, cr_level, max_output_level
):
    """Compute a gaintable for a given audiogram.

    Replaces MATLAB GUI interface of original OpenMHA software for
    gaintable_camfit_compr table calculation. Assumes two channels and
    that audiogram frequencies are identical at two ears.

    Args:
        audiogram (Audiogram): the audiogram for which to compute the gain table
        audf (list): audiogram frequencies for fitting
        noisegatelevels (ndarray): compression threshold levels for each frequency band
        noisegateslope (int): determines slope of gains below compression threshold
        cr_level (int): overall input level in dB for calculation of compression ratios
        max_output_level (int): maximum output level in dB

    Returns:
        dict: dim ndarray of gains

    """
    # Initialise parameters
    num_channels = 2  # only configured for two channels
    sFitmodel = {}

    # Fixed centre frequencies for amplification bands
    sFitmodel["frequencies"] = [
        177,
        297,
        500,
        841,
        1414,
        2378,
        4.0000e03,
        6.7270e03,
        11314,
    ]
    sFitmodel["edge_frequencies"] = [
        1.0000e-08,
        229.2793,
        385.3570,
        648.4597,
        1.0905e03,
        1.8337e03,
        3.0842e03,
        5.1873e03,
        8.7241e03,
        10000001,
    ]

    # Input levels in SPL for which to compute the gains
    sFitmodel["levels"] = np.arange(-10, 110 + 1)
    sFitmodel["channels"] = num_channels
    sFitmodel["side"] = "lr"

    # Calculate gains and compression ratios
    output = gainrule_camfit_compr(
        audiogram,
        sFitmodel,
        noisegatelevels,
        noisegateslope,
        cr_level,
        max_output_level,
    )

    sGt = {}
    sGt["sGt_uncorr"] = copy.deepcopy(output["sGt"])  # sGt without noisegate correction

    output = multifit_apply_noisegate(
        output["sGt"],
        sFitmodel["frequencies"],
        sFitmodel["levels"],
        output["noisegatelevel"],
        output["noisegateslope"],
    )

    # Reshape sGt here to suit cfg file input
    sGt["sGt"] = np.transpose(np.reshape(output["sGt"], (121, 18), order="F"))

    sGt["noisegatelevel"] = output["noisegatelevel"]
    sGt["noisegateslope"] = output["noisegateslope"]
    sGt["frequencies"] = sFitmodel["frequencies"]
    sGt["levels"] = sFitmodel["levels"]
    sGt["channels"] = num_channels

    return sGt


def format_gaintable(gaintable, noisegate_corr=True):
    """
    Format gaintable for insertion into cfg file as long string.

    Args:
        gaintable (ndarray): The gaintable to format
        noisegate_corr (boolean, optional): apply noisegate correction or do not
            (default: True)

    Returns:
        str: gaintable formatted for insertion into OpenMHA
            cfg file

    """
    if noisegate_corr:
        sGt = gaintable["sGt"]
    else:
        logging.warning("Noise gate correction not being applied to gain table.")
        sGt = gaintable["sGt_uncorr"]

    # Do not apply gains to 9th and 18th row
    logging.info("No application of gains to 9th and 18th row of table.")
    sGt[[8, -1], :] = 0

    # Format for inclusion in cfg file
    v = "["
    for k in range(0, np.shape(sGt)[0]):
        v += f"{sGt[k, :]};"
    v += "]"
    v = v.replace("\n", "")
    v = re.sub(" +", " ", v)  # remove extra white spaces
    # v = re.sub(". ", " ", v)  # remove point after integers

    formatted_sGt = v

    return formatted_sGt


def multifit_apply_noisegate(
    sGt, sFit_model_frequencies, sFit_model_levels, noisegate_level, noisegate_slope
):
    """Apply noisegate.

    Based on OpenMHA subfunction of libmultifit.m.

    Args:
        sGt (ndarray): gain array
        sFit_model_frequencies (list): FFT filterbank frequencies
        sFit_model_levels (ndarray): levels at which to calculate gains
        noisegate_level (ndarray): chosen compression threshold
        noisegate_slope (ndarray): determines slope below compression threshold

    Returns:
        ndarray: Noise signal

    """

    for i in [0, 1]:
        for kf in range(len(sFit_model_frequencies)):
            gain_noisegate = interp1d(
                sFit_model_levels, sGt[:, kf, i], fill_value="extrapolate"
            )(noisegate_level[kf, i])
            idx = [
                i for i, x in enumerate(sFit_model_levels < noisegate_level[kf, i]) if x
            ]
            sGt[idx, kf, i] = (
                sFit_model_levels[idx] - noisegate_level[kf, i]
            ) * noisegate_slope[kf, i] + gain_noisegate

    output = {}
    output["sGt"] = {}
    output["sGt"] = sGt
    output["noisegatelevel"] = noisegate_level
    output["noisegateslope"] = noisegate_slope

    return output
