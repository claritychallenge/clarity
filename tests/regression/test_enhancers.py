import json
import tempfile

import numpy as np
import torch

from clarity.enhancer.dsp import filter
from clarity.enhancer.gha.audiogram import Audiogram
from clarity.enhancer.gha.gha_interface import GHAHearingAid as gha
from clarity.enhancer.gha.gha_utils import format_gaintable, get_gaintable

gha_params = {  # hyperparameters for GHA Hearing Aid, BE CAREFUL if making changes
    "fs": 44100,
    "ahr": 20,
    "audf": None,
    "cfg_file": "prerelease_combination3_smooth",
    "noisegatelevels": None,
    "noisegateslope": 0,
    "cr_level": 0,
    "max_output_level": 100,
    "equiv0dBSPL": 100,
    "test_nbits": 16,
}


def test_dsp_filter(regtest):
    amfir = filter.AudiometricFIR(sr=44100, nfir=220, device="cpu")
    torch.manual_seed(0)
    signal = torch.rand(10, dtype=torch.float)
    signal = torch.reshape(signal, (1, 1, -1))
    output = amfir(signal.to(amfir.device))
    output = np.round(output.detach().cpu().numpy(), 4)
    regtest.write(f"signal output: \n{output}\n")


def test_gha_audiogram(regtest):
    cfs = np.array([250, 500, 1000, 2000, 4000, 8000])
    for i in [1, 10, 30, 40]:
        np.random.seed(0)
        levels_l = np.round(np.log10(np.random.rand(6) / 20) * (-i), 0)
        levels_r = np.round(np.log10(np.random.rand(6) / 20) * (-i), 0)
        ag = Audiogram(levels_l, levels_r, cfs)
        regtest.write(
            f"Audiogram original: \n{ag.cfs}\n{ag.levels_l}\n{ag.levels_r}\n{ag.severity}\n"
        )
        ag = ag.select_subset_of_cfs(np.array([500, 1000, 2000]))
        regtest.write(
            f"Audiogram new: \n{ag.cfs}\n{ag.levels_l}\n{ag.levels_r}\n{ag.severity}\n"
        )


def test_GHA_inputs(regtest):

    enhancer = gha(**gha_params)

    infile_names = [
        f"tests/test_data/scenes/S06001_mix_CH{ch}.wav" for ch in range(1, 4)
    ]
    fd_merged, merged_filename = tempfile.mkstemp(
        prefix="clarity-merged-", suffix=".wav"
    )
    enhancer.create_HA_inputs(infile_names, merged_filename)

    signal = enhancer.read_signal(merged_filename)
    np.set_printoptions(threshold=100)
    regtest.write(f"signal output: \n{signal[:,0]}\n")


def test_GHA_config(regtest):

    enhancer = gha(**gha_params)
    listeners = json.load(open("tests/test_data/metadata/listeners.json"))
    listener = listeners["L0001"]
    cfs = np.array(listener["audiogram_cfs"], dtype="int")
    audiogram_left = np.array(listener["audiogram_levels_l"])
    audiogram_right = np.array(listener["audiogram_levels_r"])
    audiogram = Audiogram(cfs=cfs, levels_l=audiogram_left, levels_r=audiogram_right)

    cfg_template = f"tests/test_data/openMHA/{enhancer.cfg_file}_template.cfg"

    gaintable = get_gaintable(
        audiogram,
        enhancer.noisegatelevels,
        enhancer.noisegateslope,
        enhancer.cr_level,
        enhancer.max_output_level,
    )
    formatted_sGt = format_gaintable(gaintable, noisegate_corr=True)

    config = enhancer.create_configured_cfgfile(
        "input", "output", formatted_sGt, cfg_template
    )
    regtest.write(f"config output: \n{config}\n")
