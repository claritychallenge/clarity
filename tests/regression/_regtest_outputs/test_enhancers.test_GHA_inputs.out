signal output:
[0. 0. 0. ... 0. 0. 0.]
